### PR TITLE
Exclude pod if istio-init container is present

### PR DESCRIPTION
**Pull request by abhide**
Issue: https://github.com/istio/istio/issues/18021

### CODE REVIEW

File: cmd/istio-cni/kubernetes.go [hunk 1]
```diff
@@ -91,3 +92,17 @@ func getK8sPodInfo(client *kubernetes.Clientset, podName, podNamespace string) (
 
 	return containers, pod.Labels, pod.Annotations, ports, nil
 }
+
+// getK8sInitContainers returns init containers of a POD
+func getK8sPodInitContainers(client *kubernetes.Clientset, podName, podNamespace string)(
+	initContainers map[string]struct{}, err error) {
+	pod, err := client.CoreV1().Pods(podNamespace).Get(podName, metav1.GetOptions{})
```
Comment (rlenglet) [inline cmd/istio-cni/kubernetes.go]: Avoid doing another k8s API call.
Instead, modify the function above to return the info we need.

File: cmd/istio-cni/main.go [hunk 2]
```diff
@@ -161,13 +163,22 @@ func cmdAdd(args *skel.CmdArgs) error {
...
+			return err
+		}
+		log.Debug("Created Kubernetes client",
+			zap.Reflect("client", client))
+		// Check if istio-init container is present; in that case exclude pod
+		initContainersMap, err := getKubePodInitContainers(client, string(k8sArgs.K8S_POD_NAME), string(k8sArgs.K8S_POD_NAMESPACE))
+		if err != nil {
+			log.Warnf("Error getting Pod data %v", err)
+		}
+		if _, present := initContainersMap[ISTIOINIT]; present {
```
Comment (rlenglet) [inline cmd/istio-cni/main.go]: 👍 This is the right logic.

File: cmd/istio-cni/kubernetes.go (hunk 1 above)
Comment (abhide) [reply cmd/istio-cni/kubernetes.go]: Done

File: cmd/istio-cni/main.go [hunk 3]
```diff
@@ -166,12 +168,18 @@ func cmdAdd(args *skel.CmdArgs) error {
...
-			containers, _, annotations, ports, k8sErr := getKubePodInfo(client, string(k8sArgs.K8S_POD_NAME), string(k8sArgs.K8S_POD_NAMESPACE))
+			log.Debug("Created Kubernetes client", zap.Reflect("client", client))
+			containers, initContainersMap, _, annotations, ports, k8sErr := getKubePodInfo(client, string(k8sArgs.K8S_POD_NAME), string(k8sArgs.K8S_POD_NAMESPACE))
 			if k8sErr != nil {
 				log.Warnf("Error getting Pod data %v", k8sErr)
 			}
+
+			// Check if istio-init container is present; in that case exclude pod
+			if _, present := initContainersMap[ISTIOINIT]; present {
+				log.Infof("Found istio-init container for pod: %s; namespace: %s", string(k8sArgs.K8S_POD_NAME), string(k8sArgs.K8S_POD_NAMESPACE))
```
Comment (rlenglet) [inline cmd/istio-cni/main.go]: Make it clear in the log that the pod is excluded, following a pattern similar to the one below:
```suggestion
				log.Info("Pod excluded due to being already injected with istio-init container",
					zap.String("pod", k8sArgs.K8S_POD_NAME),
					zap.String("namespace", k8sArgs.K8S_POD_NAMESPACE))
```

Comment (abhide) [reply cmd/istio-cni/main.go]: done

**Comment by rlenglet** [comment]
@abhide it would be nice to have a similar change in `release-1.4`. This can't be cherry-picked automatically, because I believe there are significant changes in `master`, incl. a change of logging libraries, etc.

**Comment by abhide** [comment]
Thanks @rlenglet I will cherry-pick; fix diffs and post a PR for release-1.4

**Pull request merged**